fission-web-server/library/Fission/Web/Server/Types.hs

```diff
@@ -397,15 +325,53 @@ instance IPFS.MonadLocalIPFS Server where
 
 instance IPFS.MonadRemoteIPFS Server where
   runRemote query = do
-    peerID       <- head <$> asks ipfsRemotePeers
-    IPFS.URL url <- asks ipfsURL
-    manager      <- asks httpManager
+    clusterURLs       <- asks ipfsURLs
+    IPFS.Timeout secs <- asks ipfsTimeout
+    manager           <- asks httpManager
+
+    logDebug @Text "Running IPFS request across cluster"
+    requests <- forM clusterURLs \(IPFS.URL url) ->
+      Process.asyncFor (Seconds (Unity secs)) do
+        runClientM query $ mkClientEnv manager url
 
-    Peer.connectRetry peerID 25 >>= \case
-      Left  err -> logDebug $ "Unable to connect: " <> textDisplay err
-      Right _   -> return ()
+    liftIO $ untilDone requests
 
-    liftIO . runClientM query $ mkClientEnv manager url
+    where
+      untilDone ::
+           NonEmpty (Async (Either Process.TimedOut (Either ClientError a)))
+        -> IO (Either ClientError a)
+      untilDone asyncRefs = do
+        results <- sequence (getStatus <$> asyncRefs)
+        case Process.progress results of
+          Process.Failed err -> return $ Left err
+          Process.InProgress -> untilDone asyncRefs
+          Process.Success a  -> return $ Right a
+      -- NOTE letting GHC's scheduler handle the frequency of retries.
+      -- This should be pretty lightweight, since we're just waiting for processes to return.
+      -- That said, if we go to 99% CPU or something, this is probably a good place to look 😉
```

Comment:
Per the note: we're not manually delaying the threads. Since we're asking the `Async` container if it's done, that's extremely lightweight, and I believe that the GHC scheduler is pretty smart about this.

That said, if we find that this is just spinning like wild we can add a 100ms delay to the `InProgress` case.

Reply:
yeah was searching for a poll interval. that's super cool :ok_hand:

Reply:
Turns out that there's a function that does this already called `waitAny`. It's a bit more sophisticated under the hood as well, and definitely has special stuff in the runtime. Switched to that and ripped our all of this `Process.Success` stuff.